httpx/middleware/auth.py

```diff
@@ -0,0 +1,200 @@
+
+        format_args = {
+            "username": self.username,
+            "realm": challenge.realm,
+            "nonce": challenge.nonce,
+            "uri": path,
+            "response": digest(b":".join((HA1, key_digest))),
+            "algorithm": challenge.algorithm.encode(),
+        }
+        if challenge.opaque:
+            format_args["opaque"] = challenge.opaque
+        if qop:
+            format_args["qop"] = b"auth"
+            format_args["nc"] = nc_value
+            format_args["cnonce"] = cnonce
+
+        header_value = ", ".join(
+            ['{}="{}"'.format(key, to_str(value)) for key, value in format_args.items()]
+        )
+        return "Digest " + header_value
+
+    def _get_client_nonce(self, nonce_count: int, nonce: bytes) -> bytes:
+        s = str(nonce_count).encode()
+        s += nonce
+        s += time.ctime().encode()
+        s += os.urandom(8)
+
+        return hashlib.sha1(s).hexdigest()[:16].encode()
+
+    def _get_nonce_count(self, nonce: bytes) -> typing.Tuple[int, bytes]:
+        """Returns the number of requests made with the same server provided
+        nonce value along with its 8-digit hex representation."""
+        if nonce != self._previous_nonce:
+            self._nonce_count = 1
+        else:
+            self._nonce_count += 1
+        self._previous_nonce = nonce
+        return self._nonce_count, b"%08x" % self._nonce_count
+
+
+class DigestAuthChallenge:
+    def __init__(
+        self,
+        realm: bytes,
+        nonce: bytes,
+        algorithm: str = None,
+        opaque: typing.Optional[bytes] = None,
+        qop: typing.Optional[bytes] = None,
+    ) -> None:
+        self.realm = realm
+        self.nonce = nonce
+        self.algorithm = algorithm or "MD5"
+        self.opaque = opaque
+        self.qop = qop
+
+    @classmethod
+    def from_header(cls, header: str) -> "DigestAuthChallenge":
+        """Returns a challenge from a Digest WWW-Authenticate header.
+
+        These take the form of:
+        `Digest realm="realm@host.com",qop="auth,auth-int",nonce="abc",opaque="xyz"`
+        """
+        scheme, _, fields = header.partition(" ")
+        assert scheme.lower() == "digest"
+        header_dict: typing.Dict[str, str] = {}
+        for field in fields.split(","):
+            key, value = field.strip().split("=")
+            header_dict[key] = unquote(value)
+
+        return cls.from_header_dict(header_dict)
+
+    @classmethod
+    def from_header_dict(cls, header_dict: dict) -> "DigestAuthChallenge":
+        realm = header_dict["realm"].encode()
+        nonce = header_dict["nonce"].encode()
+        qop = header_dict["qop"].encode() if "qop" in header_dict else None
+        opaque = header_dict["opaque"].encode() if "opaque" in header_dict else None
+        algorithm = header_dict.get("algorithm")
+        return cls(
+            realm=realm, nonce=nonce, qop=qop, opaque=opaque, algorithm=algorithm
+        )
```

Comment:
I have the feeling we’re over complicating this part a bit — a plain namedtuple and a function that builds it out of parsed data would be enough, in particular since we’re not really using these two class methods anywhere else. But I think the code is understandable enough, so perhaps you can call me on nitpicking. :)

Reply:
Not sure, I don't particularly like namedtuples (might be `attrs` bias) and I believe we cannot use dataclasses so I went for a class since we're supporting 3.6.

Is there anything particularly complicated about this approach?

Reply:
> Is there anything particularly complicated about this approach?

Not really — I’m fine with the current approach. :) (Though I don’t see what’s not to like about namedtuples, in particular when using the typing.NamedTuple style.)

Reply:
I mostly think they obscure what would otherwise be a lovely simple class 😄 

https://www.attrs.org/en/stable/why.html#namedtuples sums other issues as well, though not directly relevant here

Reply:
Cool, thanks for sharing the context on why attrs thinks namedtuples are bad. :)

I think they do have their use as a way of structuring/type-hinting immutable return values — as is properly noted in one of the footnotes. I don’t like the collections version of it (the one displayed in the attrs docs), but the type hint-capable version in the typing module is pretty neat for this use case.